### PR TITLE
add check for unbalanced square brackets

### LINKED ISSUE #73: Mismatched parentheses detection needs expanding to square brackets
What it says on the tin!

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
Resolves #73.

Note that I moved all unmatched parens/brackets tests to `test_calc.py` and out of FormulaGrader. Previously, i understand placing that test in `test_formulagrader.py` in order to make sure that the correct error messages actual make it through to the student. Not, I'd rather put tests closer to their source code and test that they raise `StudentFacingError`s.

### CODE REVIEW

**Comment by jolyonb** [comment]
Ok. At some point, I'm going to go through the entire apparatus and figure out just how our error handling is working...

File: mitxgraders/helpers/calc.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ class UndefinedFunction(CalcError):
     """
     pass
 
-class UnmatchedParentheses(CalcError):
+class UnbalancedBrackets(CalcError):
     """
     Indicate when a student's input has unmatched parentheses.
```
Comment (jolyonb) [inline mitxgraders/helpers/calc.py]: unmatched parentheses -> unbalanced brackets

**Comment by ChristopherChudzicki** [comment]
I was planning to tackle #75 next, but you can do it if you want :)

**Comment by jolyonb** [comment]
Nope, by all means, all yours!

**Comment by ChristopherChudzicki** [comment]
Well, I completely overhauled the bracket balancing and the error messages. Now students see things like this:

screen shot 2018-07-13 at 1 12 37 am

Also addresses #82

**Comment by jolyonb** [comment]
This is really pretty. Good job!

**Comment by jolyonb** [comment]
Resolves #82.

**Comment by ChristopherChudzicki** [comment]
Fixed the typos :) 

> You obviously had fun writing this. It's very cute :-) Two minor typos.

It was fun :) and the deeply nested for / if's in my earlier approach were driving me nuts.

**Comment by jolyonb** [comment]
A FILO stack is definitely the way to do this. You made it much prettier than I would've though!

**Pull request merged**